File: competitive_programming/leetcode/01_python/202_happy_number.py
"""LeetCode 202 - Happy Number."""
def happy_number(n):
    seen = set()
    while n != 1 and n not in seen:
        seen.add(n)
        n = sum(int(digit) ** 2 for digit in str(n))
    return n == 1


if __name__ == "__main__":
    assert happy_number(19) is True
    assert happy_number(2) is False
    print("202 happy number ok")

File: competitive_programming/leetcode/03_cpp/202_happy_number.cpp
#include <algorithm>
#include <cassert>
#include <cmath>
#include <cstdint>
#include <iostream>
#include <set>
#include <string>
#include <vector>

bool happyNumber(int n) {
    std::set<int> seen;
    while (n != 1 && seen.find(n) == seen.end()) {
        seen.insert(n);
        int next = 0;
        while (n > 0) { int digit = n % 10; next += digit * digit; n /= 10; }
        n = next;
    }
    return n == 1;
}

int main() {
    assert(happyNumber(19) && !happyNumber(2));
    std::cout << "202 happy number ok\n";
    return 0;
}
